### ISSUE TITLE
 info = cnn_train_dag(..) is correct?

### ISSUE THREAD

**Issue by alevillard** [vlfeat/matconvnet#554]
Hi, my training code is kind like this (I will skip some parts), so I have two output variable in my training function: net and info. Now, I've seen that both of them after a training process are such DagNN variable, and in the function cnn_train_dag, it has two outputs in the code: function [net,stats] = cnn_train_dag(net, imdb, getBatch, varargin). But when I use this cnn_train_dag in the training example code I have only info=cnn_train_dag(net, imdb, @(i,b) getBatchDisk(bopts,i,b), opts.train, 'val', find(imdb.images.set == 2)); So a single output from cnn_train_dag instead of two. This what it means? that my dagNN network is info? or net as well.  Thank you for reply, and sorry for this (stupid) question. 

function [net, info] = vggD_train(imdb, expDir, varargin)

% some common options
	opts.train.batchSize = 30;
	opts.train.numEpochs = 35 ;
	opts.train.continue = true ;
	opts.train.gpus = [1] ;
	opts.train.learningRate = [1e-2*ones(1, 25),  1e-3*ones(1, 15)];
	opts.train.weightDecay = 5e-4;
	opts.train.momentum = 0.5; 
	opts.train.expDir = expDir;
	opts.train.numSubBatches = 1;
	% getBatch options
	bopts.useGpu = numel(opts.train.gpus) >  0 ;
	opts.optMethod = 'gradient'; % ['adagrad', 'gradient']

     if(numel(varargin) > 0)
		netPre = varargin{1};
    end
% network definition!
	% MATLAB handle, passed by reference
	net = dagnn.DagNN() ;

.....network layers..

net.addLayer('prob', dagnn.SoftMax(), {'classifier'}, {'prob'}, {});
	net.addLayer('objective', dagnn.Loss('loss', 'log'), {'prob', 'label'}, {'objective'}, {});
	net.addLayer('error', dagnn.Loss('loss', 'classerror'), {'prob','label'}, 'error') ;
	% -- end of the network

    % initialization of the weights 
    if(numel(varargin) > 0)
		initNet_FineTuning(net, netPre);
        
	else
		 initNet_He(net);
	end	
   
    %train
    info = cnn_train_dag(net, imdb, @(i,b) getBatchDisk(bopts,i,b), opts.train, 'val', find(imdb.images.set == 2)) ;
end

**Comment by samadak** [comment]
Hi 
Thanks for time and your energy
I am going to initialize only  two last  layer. I  am so confused about InitNet function
I was copy the code but i faced up to error
I am appreciated

**Comment by alevillard** [comment]
Hi samadak, 
you should add this other rows of code I attach below to make work the functions InitNet_He, InitNet_Finetuning and getBatchDisk. For the init there are params you can modify by yourself to find a better convergence of the model in the training; it depends on the network. 


function initNet_FineTuning(net, netPre)
 net.initParams();
 ind = 1;
 for l=1:length(net.layers)-7
 % is a convolution layer?
 if(strcmp(class(net.layers(l).block), 'dagnn.Conv'))

 f_ind = net.layers(l).paramIndexes(1);
 b_ind = net.layers(l).paramIndexes(2);
 net.params(f_ind).value = netPre.params(f_ind).value;
 ind = ind + 1;
 net.params(b_ind).value = netPre.params(b_ind).value;
 ind = ind + 1;
 if l<= length(net.layers)-7
 net.params(f_ind).learningRate = 1;
 net.params(f_ind).weightDecay = 1;
 net.params(b_ind).learningRate = 0.5;
 net.params(b_ind).weightDecay = 1;

 else
 net.params(f_ind).learningRate = 1;
 net.params(f_ind).weightDecay = 1;
 net.params(b_ind).learningRate = 0.5;
 net.params(b_ind).weightDecay = 1;
 end
 end
 if(strcmp(class(net.layers(l).block), 'dagnn.BatchNorm'))
 f_indbn = net.layers(l).paramIndexes(1);
 b_indbn = net.layers(l).paramIndexes(2);
 m_indbn = net.layers(l).paramIndexes(3);

 net.params(f_indbn).value = netPre.params(f_indbn).value;
 %ind = ind + 1;
 net.params(b_indbn).value = netPre.params(b_indbn).value;
 %ind = ind + 1;
 net.params(m_indbn).value = netPre.params(m_indbn).value;
 %ind = ind + 1;

 net.params(f_indbn).learningRate = 1;
 net.params(f_indbn).weightDecay = 1;
 net.params(b_indbn).learningRate = 1;
 net.params(b_indbn).weightDecay = 1;
 net.params(m_indbn).learningRate = 0.01;
 net.params(m_indbn).weightDecay = 1;
 end 

 end
 for l=length(net.layers)-6:length(net.layers)
 % is a convolution layer?
 if(strcmp(class(net.layers(l).block), 'dagnn.Conv'))
 f_ind = net.layers(l).paramIndexes(1);
 b_ind = net.layers(l).paramIndexes(2);

 [h,w,in,out] = size(net.params(f_ind).value);
 he_gain = 1e-1*sqrt(2/(h*w*in)); % sqrt(1/fan_in)
 net.params(f_ind).value = he_gain*randn(size(net.params(f_ind).value),
'single');
 net.params(f_ind).learningRate = 1;
 net.params(f_ind).weightDecay = 1;
 net.params(b_ind).value = zeros(size(net.params(b_ind).value), 'single');
 net.params(b_ind).learningRate = 0.5;
 net.params(b_ind).weightDecay = 1;

 end
 end
end
function initNet_He(net)
 net.initParams();
 for l=1:length(net.layers)
 % is a convolution layer?
 if(strcmp(class(net.layers(l).block), 'dagnn.Conv'))
 f_ind = net.layers(l).paramIndexes(1);
 b_ind = net.layers(l).paramIndexes(2);
 [h,w,in,out] = size(net.params(f_ind).value);
 he_gain = 1e-2*sqrt(2/(h*w*in)); % sqrt(1/fan_in)
 net.params(f_ind).value = 2*he_gain*randn(size(net.params(f_ind).value),
'single');
 net.params(f_ind).learningRate = 1;
 net.params(f_ind).weightDecay = 1;
 net.params(b_ind).value = zeros(size(net.params(b_ind).value), 'single');
 net.params(b_ind).learningRate = 0.5;
 net.params(b_ind).weightDecay = 1;
 end
 if(strcmp(class(net.layers(l).block), 'dagnn.BatchNorm'))
 f_indbn = net.layers(l).paramIndexes(1);
 b_indbn = net.layers(l).paramIndexes(2);
 m_indbn = net.layers(l).paramIndexes(3);
 net.params(f_indbn).learningRate = 1;
 net.params(f_indbn).weightDecay = 1;
 net.params(b_indbn).learningRate = 1;
 net.params(b_indbn).weightDecay = 1;
 net.params(m_indbn).learningRate = 0.01;
 net.params(m_indbn).weightDecay = 1;
 end
 end 

end
% getBatch for IMDBs that are too big to be in RAM
function inputs = getBatchDisk(opts, imdb, batch)
 images = zeros(224, 224, 3, 'single');
for i=1:numel(batch)
 im = imread(imdb.images.filenames{batch(i)}) ;
 im_ = single(im) ; im_=imresize(im_, [224 224]);
 im_= im_ - imdb.images.data_mean;
 images(:,:,:, i) = im_;
end
clear im; clear im_;
 labels = imdb.images.labels(1,batch) ;
 if opts.useGpu > 0
 images = gpuArray(images) ;
 end
 inputs = {'input', images, 'label', labels} ;
end

**Comment by alevillard** [comment]
In the rows with he_gain, it's missed a * symbol for the multiplication, be careful

he_gain = 1e-1*sqrt(2/(hwin))
net.params(f_ind).value = he_gain*randn
he_gain = 1e-2*sqrt(2/(hwin))
2*he_gain*
